diff -ur busybox-w32.bak/include/mingw.h busybox-w32/include/mingw.h
--- busybox-w32.bak/include/mingw.h	2023-05-28 17:13:49.000000000 +0800
+++ busybox-w32/include/mingw.h	2023-07-27 19:56:23.255724600 +0800
@@ -219,7 +219,11 @@
 /*
  * strings.h
  */
+#if __clang__
+#define ffs __builtin_ffs
+#else
 int ffs(int i);
+#endif
 
 /*
  * sys/ioctl.h
diff -ur busybox-w32.bak/scripts/Makefile.build busybox-w32/scripts/Makefile.build
--- busybox-w32.bak/scripts/Makefile.build	2023-05-28 17:13:49.000000000 +0800
+++ busybox-w32/scripts/Makefile.build	2023-07-27 19:47:22.634569300 +0800
@@ -257,7 +257,7 @@
 # If the list of objects to link is empty, just create an empty built-in.o
 # -nostdlib is added to make "make LD=gcc ..." work (some people use that)
 cmd_link_o_target = $(if $(strip $(obj-y)),\
-		$(LD) -nostdlib $(ld_flags) -r -o $@ $(filter $(obj-y), $^),\
+		$(if $(filter 1,$(words $(filter-out %dummy.o,$(filter $(obj-y), $^)))),cp $(filter-out %dummy.o,$(filter $(obj-y), $^)) $@,$(LD) -nostdlib $(ld_flags) -c -o $@ $(filter $(obj-y), $^)),\
 		rm -f $@; $(AR) rcs $@)
 
 $(builtin-target): $(obj-y) FORCE
diff -ur busybox-w32.bak/scripts/trylink busybox-w32/scripts/trylink
--- busybox-w32.bak/scripts/trylink	2023-05-28 17:13:49.000000000 +0800
+++ busybox-w32/scripts/trylink	2023-07-27 20:00:11.783498600 +0800
@@ -96,7 +96,7 @@
 START_GROUP="-Wl,--start-group"
 END_GROUP="-Wl,--end-group"
 INFO_OPTS() {
-	echo "-Wl,--warn-common -Wl,-Map,$EXE.map -Wl,--verbose"
+	echo "-Wl,-Map,$EXE.map -Wl,--verbose"
 }
 
 # gold may not support --sort-common (yet)
@@ -320,7 +320,6 @@
 		$SORT_SECTION \
 		$GC_SECTIONS \
 		-L"$sharedlib_dir" -lbusybox \
-		-Wl,--warn-common \
 	|| {
 	    echo "Linking $EXE failed"
 	    cat $EXE.out

