diff --git a/win32/process.c b/win32/process.c
index 44bc1f3..3fcc6a2 100644
--- a/win32/process.c
+++ b/win32/process.c
@@ -554,6 +554,7 @@ mingw_execve(const char *cmd, char *const *argv, char *const *envp)
 {
 	intptr_t ret = mingw_spawn_interpreter(P_NOWAIT, cmd, argv, envp, 0);
 
+#if defined(__MINGW64_VERSION_MAJOR) && (__MINGW64_VERSION_MAJOR >= 11)
 	if (ret == -1 && GetLastError() == ERROR_ELEVATION_REQUIRED) {
 		// Command exists but failed because it wants elevated privileges.
 		// Try again using ShellExecuteEx().
@@ -562,6 +563,7 @@ mingw_execve(const char *cmd, char *const *argv, char *const *envp)
 		if (GetLastError())
 			exit(1);
 	}
+#endif
 
 	if (ret != -1)
 		wait_for_child((HANDLE)ret, cmd);
